PULL_REQUEST_TEMPLATE.md

```diff
@@ -1,12 +1,42 @@
-<!-- Your title above should be a short description of what
-was changed. Do not include the issue number in the title. -->
+<!--
+Thank you so much for your PR!  To help us review your contribution, please
+consider the following points:
+
+- Your title above should be a short description of what was changed. Do not
+  include the issue number in the title.
+
+- Development information is available at
+  https://github.com/sympy/sympy/wiki/Development-workflow.
+
+- Do not create the PR out of master, but out of a separate branch.
+
+- Try to keep the number of commits to a minimum. Use rebase and
+  force pushing in git to avoid adding unnecessary commits.
```

Comment:
This is problematic if you are trying to help someone by adding commits. They will push changed history to the public. I don't think we should recommend force pushing.

Reply:
Is that a common practice though? It doesn't say you have to use a single commit, but 50 commits for 15 lines, where 20 have a commit message of the type "debugging", is not something I believe contribute to the code base.

Reply:
I think when you follow the pull request model then rebasing is fine. The changed history is only problematic for long-lived shared branches.

Reply:
If you want to address the 20 commits "debugging", then we should state something more like "Minimize the number of commits by squashing your latest commits before pushing to your branch on GIthub."

Reply:
That is basically what I am looking for, yes, and do not really see any major difference. It will still involve force-pushing though (since many do not run complete test-suite locally, which is sort of obvious because if they did run them locally there is no need to commit before it worked).

Also, I believe that rebasing is still better than merging in most situations. For example, squashing merge commits is maybe not the best idea.

But I am definitely open to rephrase it in a satisfactory way.

(As a reference, some well-known repos do, basically, not allow merge commits at all.)

Reply:
>  It will still involve force-pushing though

No, it will not require force pushing. As long as you don't squash any commits that are already pushed to Github, then you can do all the squashing you want on your on computer.

Reply:
To be honest, why do we even care if there are a bunch of "debugging" commits. It doesn't really affect anything down the road. New contributors will do this, but will learn not to later on. I don't think there is an actual technical problem we are solving by not allowing a bunch of tiny commits in the tree. Yes, it would be extra nice if everyone had very disciplined git practices, but I think the reality is that most people will not, and that is ok. I'd take merge commits + numerous ill-formatted tiny commits over having to coach people through rebasing all the time. Rebasing anything that isn't a trivial hop up a linear divergent branch, is often a pain in the ass. I avoid rebasing myself because I hate trying to work it on on the command line. It's so much simpler to just merge in the branch you are tracking.

Reply:
Here is my proposal:

Try to keep the number of commits to a minimum in your pull request branch. Avoid committing many very small changes, but if you do, you can use Git's rebase command locally to squash all of the commits since your last push. Avoid force pushing to the branch of your pull request, because that rewrites the public history and makes it difficult to collaborate with other SymPy developers.

Reply:
Just a suggestion here - can't we always do squash and merge option while merging the PR, this way members would be able to provide decent commit message and let the contributor have small commits?

Reply:
> New contributors will do this, but will learn not to later on.

Why shouldn't they learn this immediately?

They would learn a lot faster if it was a necessary part of getting a PR merged. Alternatively if it's too much for a new contributor to be expected to do this then the person merging could do it as suggested in #18548.

Either way if we don't try to enforce a clean history then we will continue to have a log full of gibberish:
https://github.com/sympy/sympy/commits/master

> Avoid force pushing to the branch of your pull request, because that rewrites the public history and makes it difficult to collaborate with other SymPy developers.

This is not really true in the pull request development model. It applies to long-lived shared branches. A long time ago SymPy had those but now there is only one long-lived shared branch: master. Clearly force-pushing master is bad. Every other branch though is a PR which is owned by a single person and should be fine to force-push. We should encourage using rebase and force-push to improve the history.

The only caveat I would have though is that I would prefer any squashing to happen at the end just before merging. While a PR is still in development it can be useful to be able to see the incremental steps. Most PRs with many commits in sympy though do not have an internal history that is useful in retrospect.

Reply:
Yes, that is possible, but would need some work to build that into SymPy's culture and processes. We've always not been too strict about git commits, messages, and git-onic (like pythonic) practices and we've always encouraged merging and merge commits as a way to lower the overhead for making contributions.

Reply:
My big issue with this is that rebasing/force pushing is a major complexity jump for new git users. Tt still trips me up from time to time, and I would consider myself to be a git expert (to be fair, I very rarely rebase, but that's precisely because I know that it is so complicated and error prone).  

It's even worse when they mess it up and have to fix things. That's why I generally prefer to not bother new contributors with such things, unless their history is particularly egregious. Note that if they have the "allow contributors to push" setting enabled (which it is by default) you can always do this for them right before merging. 

I also personally feel that squashing results in bad committing habits with lots of "fixup" and similarly useless commit messages. It's better to get in the habit of making well structured diffs and writing good commit messages the first time. 

That's why I've never really been in favor of recommending people to rebase or squash their branches. If they want to do it fine. Rebasing without squashing is also quite useless in my opinion. You're much better off just using a merge commit.

Reply:
> Note that if they have the "allow contributors to push" setting enabled (which it is by default) you can always do this for them right before merging.

Note also that there is a button in GitHub that can be used to do this directly when merging a PR (https://github.com/sympy/sympy/issues/18548). A bunch of people have given the thumbs up but you and Jason are the only people who have objected to enabling that button in the sympy repo.

The options are basically:

1. Enable squash merging.
2. Ask contributors to squash their commits.
3. Have a complete mess of a git history.

Currently we have a mix of 2 and 3. Your 4th option that the person merging the PR should manually squash the commits is a non-option to me: if that's something that should be done why is the button that makes it easy disabled?

I have manually squashed someone's commits very occasionally but 99% of the time if the commits need to be squashed I will ask the contributor to do it because I want them to understand why it needs doing and how to do it and to avoid the same situation in future.

Reply:
I've never used that button, it's relatively new (I guess). Is it somehow smart enough to solve all the rebase merge conflict problems you get on the command line? If it is somehow smarter than the command line and squash at the end before merge just always works, then I'm not necessarily opposed to enabling it. My reasons for discouraging rebasing is because it is often such a painful process and getting lost in the rebase merge conflict scenario is way more confusing and difficult to get out of than a simple merge conflict when merging in a branch you are tracking. Maybe github's button has magic that relieves you of these issues.

But I still don't think your point 3 is actually an issue. The only real issue that new contributors can cause in the git tree is if they do things like delete directories of code, commit, and then bring them back. Or modify every file in the code base by changing its timestamp or something. Those things do actually cause a blip in the history that makes it hard to use blame. But a bunch of "debugging" commits don't actually cause any technical issues. Sure the log is longer and we may have to scan through more commit message, but that is only an issue for those desiring some idealized git commit history.

Reply:
> Why shouldn't they learn this immediately?

Because every little barrier we include, adds up and can discourage new people from contributing. The barrier for entry should be as low as we can possibly make it.

Reply:
> Either way if we don't try to enforce a clean history then we will continue to have a log full of gibberish:

Why do you care if it is gibberish or not? Is there a technical issue that making it non-gibberish solves?

Reply:
> No, it will not require force pushing. As long as you don't squash any commits that are already pushed to Github, then you can do all the squashing you want on your on computer.

Yes, it will. Squashing locally is not a solution when tests are not executed locally.

> Why do you care if it is gibberish or not?

Why do you write commit messages?